### ISSUE TITLE
Please include "Subject Alternative Name=localhost" to s3mock.jks certificate

### ISSUE THREAD

**Issue by vlsi** [adobe/S3Mock#228]
The mock users should be able to access mock server with a regular TLS connection and proper hostname verification.

If I pull `s3mock.jks` and use it as a truststore, then I get the following error:

```
Unable to execute HTTP request: Certificate for <localhost> doesn't match any of the subject alternative names: []
```

Note: I can't use `s3mock`-provided S3 client. I want the application under test to run as usual, and the only change I do is I configure `endpoint` to be `localhost:port` and configure `enablePathStyleAccess`.

---

PS It would be great if there was documentation on the way to use a custom-generated certificate for s3mock TLS server.

I managed to configure it as follows:
1. Create self-signed certificate with localhost subject alternative name

        keytool -genkey -keyalg RSA -alias selfsigned -keystore keystore.jks -storepass password  -keysize 2048 -ext san=dns:localhost

2. Configure s3mock to use it. I had to add process-level properties. It would be great to be able to configure mock server properties on a per-test basis. It looks like `@ExtendWith(S3MockExtension::class)` does not support that.

                server.ssl.key-store=classpath:test_keystore.jks
                server.ssl.key-alias=selfsigned
                server.ssl.key-store-password=password
                server.ssl.key-password=password

3. Configure client Java to use the same keystore as truststore:

                javax.net.ssl.trustStore=path_to/test_keystore.jks
                systemProp.javax.net.ssl.trustStorePassword=password

Now my test works without "trust any certificate" workarounds.

**Comment by vlsi** [comment]
This is related to https://github.com/adobe/S3Mock/pull/123

**Comment by vlsi** [comment]
https://nip.io/ might help here as well.

127-0-0-1.nip.io resolves to 127.0.0.1, and us-west-1.127-0-0-1.nip.io resolves to 127.0.0.1 as well.

That would enable users to skip `enablePathStyleAccess` and connect to s3mock via regular `us-west-1.<endpoint>`.

Then, the flow could be as follows:

1) S3Mock could include a root certificate like it has currently. The clients might add the certificate to their trust store during the test.
2) S3Mock could create a temporary certificate before it starts the server. The temporary certificate should be signed with the root certificate, and the temporary would include `Subject Alternative Name` == (`localhost`, `127.0.0.1`, `127-0-0-1.nip.io`, `us-west-1.127-0-0-1.nip.io`, ...)
3) S3MockStarter (or something like that) should expose `getEndpoint` which could be something like `127-0-0-1.nip.io:port`.

Note: there might be a helper method from S3Mock to add S3Mock's root certificate to the default system: https://stackoverflow.com/a/62586564/1261287
In other words, either the user specifies root cert via `javax.net.ssl.trustStore` or they call something like `S3MockApplication.installS3MockRootCertificateToSystemTrustStore()` before they access S3 API.

WDYT?

**Comment by vlsi** [comment]
It turns out pure Java does not have API to sign certificates, so BouncyCastle should be used to do that.
The signing is not that complicated: https://stackoverflow.com/a/7366757/1261287

Do you think it is ok to add BouncyCastle as a runtime dependency?

**Comment by afranken** [comment]
Just re-generated the certificate / keystore:

```
keytool -genkey -keyalg RSA -alias selfsigned -keystore s3mock.jks -storepass password  -keysize 2048 -ext san=dns:localhost -validity 365000
What is your first and last name?
  [Unknown]:  Adobe S3Mock
What is the name of your organizational unit?
  [Unknown]:  S3Mock
What is the name of your organization?
  [Unknown]:  S3Mock
What is the name of your City or Locality?
  [Unknown]:  Hamburg
What is the name of your State or Province?
  [Unknown]:  Hamburg
What is the two-letter country code for this unit?
  [Unknown]:  DE
Is CN=Adobe S3Mock, OU=S3Mock, O=S3Mock, L=Hamburg, ST=Hamburg, C=DE correct?
  [no]:  yes

Generating 2.048 bit RSA key pair and self-signed certificate (SHA256withRSA) with a validity of 365.000 days
	for: CN=Adobe S3Mock, OU=S3Mock, O=S3Mock, L=Hamburg, ST=Hamburg, C=DE
```
